winch/codegen/src/visitor.rs

```diff
@@ -1189,13 +1189,39 @@ where
             self.context.pop_to_reg(self.masm, None)
         };
 
-        self.masm.branch(
-            IntCmpKind::Ne,
-            top.reg.into(),
-            top.reg.into(),
-            *frame.label(),
-            OperandSize::S32,
-        );
+        // Emit instructions to balance the machine stack if the frame has
+        // a different offset.
+        let current_sp_offset = self.masm.sp_offset();
+        let (_, frame_sp_offset) = frame.base_stack_len_and_sp();
+        if current_sp_offset == frame_sp_offset {
```

Comment:
Minor note here: instead of checking for equality, can we check for `current_sp_offset > frame_sp_offset`? We have checks everywhere else that guarantee that the current sp should never be less than the target SP; doing so will also align with how `ensure_sp_for_jump` is built and it also conveys a clearer message that the stack will be manipulated.

Reply:
Should `SPOffset` implement `Ord`? I can call `.as_u32()` on both sides, but I _think_ it would be reasonable to be able to compare them directly.

Reply:
Yeah, I think it's fine for it to implement `Ord`.

Reply:
We can do it in a follow-up PR too.